switch_locale_context: Add aTHX

khwilliamson opened the pull request:
This fixes GH #21040

Instead of a dTHX, this passes aTHX automatically, and skips calling this function if there is no valid context.

It moves that decision into the macro itself, avoiding some #ifdef directives.

mauke commented on locale.c:
```diff
@@ -7219,10 +7219,7 @@ Perl_switch_locale_context()
 
     /* Can't use pTHX, because we may be called from a place where that
```
Why does this comment say "Can't use pTHX" if the function now uses `pTHX`?

khwilliamson replied:
Because I made a mistake; now fixed

khwilliamson committed:
switch_locale_context: Add aTHX

This fixes GH #21040

Instead of a dTHX, this passes aTHX automatically, and skips calling
this function if there is no valid context.

It moves that decision into the macro itself.

khwilliamson commented:
This doesn't work on Windows
